docs/new-feature-tutorial.rst

```diff
@@ -2,84 +2,178 @@
-that actually updates the relevant field.  In this case, `do_set_realm_invite_by_admins_only`
-is a function that actually updates the field in the database, and sends
-an event announcing that this change has been made.
-
-You then need update the `fetch_initial_state_data` and `apply_events` functions
-in `zerver/lib/actions.py` to update the state based on the event you just created.
-In this case, we add a line
-
-::
-
-  state['realm_invite_by_admins_only'] = user_profile.realm.invite_by_admins_only`
-
-to the `fetch_initial_state_data` function.  The `apply_events` function
-doesn't need to be updated since
-
-::
-
-   elif event['type'] == 'realm':
-       field = 'realm_' + event['property']
-       state[field] = event['value']
+**Database interaction:** Add any necessary code for updating and interacting
+with the database in ``zerver/lib/actions.py``. It should update the database and 
+send an event announcing the change.
 
-already took care of our event.
+**Application state:** Modify the ``fetch_initial_state_data`` and ``apply_events`` 
+functions in ``zerver/lib/actions.py`` to update the state based on the event you 
+just created.
 
-Then update `zerver/views/__init__.py` to actually call your function.
-In the dictionary which sets the javascript `page_params` dictionary,
-add a value for your feature.
+**Backend rendering:** Make any other modifications to the backend required for 
+your change. This could include the [xxx] file, which controls the generation of 
+various pages served (along with the Django templates).
 
-::
+**Testing:** At the very least, add a test of your event data flowing through the system in
+``test_events.py``.
 
-   realm_invite_by_admins_only = register_ret['realm_invite_by_admins_only']
 
-Perhaps your new option controls some other backend rendering: in our case
-we test for this option in the `home` method for adding a variable to the response.
-The functions in this file control the generation of various pages served
-(along with the Django templates).
-Our new feature also shows up in the administration tab (as a checkbox),
-so we need to update the `update_realm` function.
+Frontend changes
+----------------
 
+**JavaScript:** Zulip's JavaScript is located in the directory ``static/js/``. 
+The exact files you may need to change depend on your feature. If you've added a 
+new event that is sent to clients, be sure to add a handler for it to
+``static/js/server_events.js``.
 
-Finally, add tests for your backend changes; at the very least you
-should add a test of your event data flowing through the system in
-``test_events.py``.
+**CSS:** The primary CSS file is ``static/styles/zulip.css``. If your new 
+feature requires UI changes, you may need to add additional CSS to this file.
 
+**Templates:** Zulip uses Handlebars templates which are located in
+``static/templates``. Templates are precompiled as part of the build/deploy
+process.
```

Comment:
Also mention if you're changing part of the initial page `template/zerver` is where those templates are (using the Django templating system).